### PR TITLE
better error message for failed Github Authentication requests

### LINKED ISSUE #9575: Vague error message with Github authentication
### Describe the bug

When a user has a bad client_id or client_secret, the following exception is returned:   `Error from RemoteAuthentication: Failed to retrieve access token.. `
### Logs

```
2021-05-28 11:20:43.1770|Default|00-512df502164b1c47be9d61b690920e3b-4832946b779fb845-00||Microsoft.AspNetCore.Mvc.ViewFeatures.ViewResultExecutor|INFO|Executing ViewResult, running view Login. 
2021-05-28 11:20:43.6856|Default|00-512df502164b1c47be9d61b690920e3b-4832946b779fb845-00||Microsoft.AspNetCore.Mvc.ViewFeatures.ViewResultExecutor|INFO|Executed ViewResult - view Login executed in 2637.6773ms. 
2021-05-28 11:20:43.6856|Default|00-512df502164b1c47be9d61b690920e3b-4832946b779fb845-00||Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker|INFO|Executed action OrchardCore.Users.Controllers.AccountController.Login (OrchardCore.Users) in 2708.9458ms 
2021-05-28 11:20:43.6856|Default|00-512df502164b1c47be9d61b690920e3b-4832946b779fb845-00||Microsoft.AspNetCore.Routing.EndpointMiddleware|INFO|Executed endpoint 'OrchardCore.Users.Controllers.AccountController.Login (OrchardCore.Users)' 
2021-05-28 11:20:43.6894|Default|00-512df502164b1c47be9d61b690920e3b-4832946b779fb845-00||Microsoft.AspNetCore.Hosting.Diagnostics|INFO|Request finished HTTP/2 GET https://localhost:5001/Login - - - 200 - text/html;+charset=utf-8 3683.7079ms 
2021-05-28 11:20:45.0939|None|00-c3c6324f709afd448fb0c265cf7c0b64-b209cdc1c637b241-00||Microsoft.AspNetCore.Hosting.Diagnostics|INFO|Request starting HTTP/2 POST https://localhost:5001/OrchardCore.Users/Account/ExternalLogin application/x-www-form-urlencoded 198 
2021-05-28 11:20:45.0939|Default|00-c3c6324f709afd448fb0c265cf7c0b64-b209cdc1c637b241-00||OrchardCore.Modules.ModularTenantRouterMiddleware|INFO|Begin Routing Request 
2021-05-28 11:20:45.0939|Default|00-c3c6324f709afd448fb0c265cf7c0b64-b209cdc1c637b241-00||Microsoft.AspNetCore.Routing.EndpointMiddleware|INFO|Executing endpoint 'OrchardCore.Users.Controllers.AccountController.ExternalLogin (OrchardCore.Users)' 
2021-05-28 11:20:45.0939|Default|00-c3c6324f709afd448fb0c265cf7c0b64-b209cdc1c637b241-00||Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker|INFO|Route matched with {area = "OrchardCore.Users", action = "ExternalLogin", controller = "Account"}. Executing controller action with signature Microsoft.AspNetCore.Mvc.IActionResult ExternalLogin(System.String, System.String) on controller OrchardCore.Users.Controllers.AccountController (OrchardCore.Users). 
2021-05-28 11:20:45.1182|Default|00-c3c6324f709afd448fb0c265cf7c0b64-b209cdc1c637b241-00||Microsoft.AspNetCore.Mvc.ChallengeResult|INFO|Executing ChallengeResult with authentication schemes (GitHub). 
2021-05-28 11:20:45.1182|Default|00-c3c6324f709afd448fb0c265cf7c0b64-b209cdc1c637b241-00||OrchardCore.GitHub.Configuration.GitHubHandler|INFO|AuthenticationScheme: GitHub was challenged. 
2021-05-28 11:20:45.1182|Default|00-c3c6324f709afd448fb0c265cf7c0b64-b209cdc1c637b241-00||Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker|INFO|Executed action OrchardCore.Users.Controllers.AccountController.ExternalLogin (OrchardCore.Users) in 11.9974ms 
2021-05-28 11:20:45.1182|Default|00-c3c6324f709afd448fb0c265cf7c0b64-b209cdc1c637b241-00||Microsoft.AspNetCore.Routing.EndpointMiddleware|INFO|Executed endpoint 'OrchardCore.Users.Controllers.AccountController.ExternalLogin (OrchardCore.Users)' 
2021-05-28 11:20:45.1182|Default|00-c3c6324f709afd448fb0c265cf7c0b64-b209cdc1c637b241-00||Microsoft.AspNetCore.Hosting.Diagnostics|INFO|Request finished HTTP/2 POST https://localhost:5001/OrchardCore.Users/Account/ExternalLogin application/x-www-form-urlencoded 198 - 302 0 - 27.3245ms 
2021-05-28 11:20:45.3009|None|00-ce1e53fe57dcc743bf276a2cb92dd5d1-a779b30ba4f78549-00||Microsoft.AspNetCore.Hosting.Diagnostics|INFO|Request starting HTTP/2 GET https://localhost:5001/signin-github?code=bf5a4d1f27cf172b9ae8&state=CfDJ8EBQ09VC_CJIqzzLIiGitF-HQZyHMpN_BADDzy5Xq-bdkX_RusVVuuGLZTp7qlDK26Jo03FiBEH_c5dYaYdJNMezv37h_elKMaczlUzZydqgtt3ZbhsWEXN_UseQ4a7ErmvH4DT70nlK5S2fiWrB3Hwso-_COkYsPYEo4QnzTenFaZcp5OmZHQVGgALjoW8qogr0XDMdZjCVCpeamX_B3gUWJ32iasz8vsEWFjxEVcdHe_B_GTApSy3bbwhibm57C8X2sbDBAuc41vJkLmJwHD7t9k8x-27QnXbtO3LiGyOl - - 
2021-05-28 11:20:45.3009|Default|00-ce1e53fe57dcc743bf276a2cb92dd5d1-a779b30ba4f78549-00||OrchardCore.Modules.ModularTenantRouterMiddleware|INFO|Begin Routing Request 
2021-05-28 11:20:45.4723|Default|00-ce1e53fe57dcc743bf276a2cb92dd5d1-a779b30ba4f78549-00||OrchardCore.GitHub.Configuration.GitHubHandler|INFO|Error from RemoteAuthentication: Failed to retrieve access token.. 
2021-05-28 11:20:45.4723|Default|00-ce1e53fe57dcc743bf276a2cb92dd5d1-a779b30ba4f78549-00||Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware|ERROR|An unhandled exception has occurred while executing the request. System.Exception: An error was encountered while handling the remote login.
 ---> System.Exception: Failed to retrieve access token.
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler`1.HandleRequestAsync()
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(HttpContext context)
   at OrchardCore.Diagnostics.DiagnosticsStartupFilter.<>c__DisplayClass3_0.<<Configure>b__1>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at OrchardCore.Modules.ModularTenantRouterMiddleware.Invoke(HttpContext httpContext)
   at OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func`2 execute, Boolean activateShell)
   at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)    at Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler`1.HandleRequestAsync()
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(HttpContext context)
   at OrchardCore.Diagnostics.DiagnosticsStartupFilter.<>c__DisplayClass3_0.<<Configure>b__1>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at OrchardCore.Modules.ModularTenantRouterMiddleware.Invoke(HttpContext httpContext)
   at OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func`2 execute, Boolean activateShell)
   at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

### PR DESCRIPTION
**Pull request by jptissot**
fix #9575

I am not sure if this should be in Orchard's code or not but it outputs a better exception message

image

### CODE REVIEW

**Comment by jptissot** [comment]
I am guessing the API changed from .net 3.1 and .net 5

**Comment by jtkech** [comment]
@jptissot 

I fixed the build, one drawback when compiling both under 3.1 and 5.0 is that under 3.1 there are some methods that we can't use, here we can't use `ReadAsStringAsync(cancellationToken)` under 3.1, only `ReadAsStringAsync()`

**Comment by jptissot** [comment]
Ah ok, I copied this code from the aspnet repo. Not sure if this fix makes sense, but it does highlight the correct error coming from the github api.

File: src/OrchardCore.Modules/OrchardCore.GitHub/Configuration/GithubHandler.cs [hunk 1]
```diff
@@ -36,5 +44,65 @@ protected override async Task<AuthenticationTicket> CreateTicketAsync(ClaimsIden
...
+                {
+                    var output = new StringBuilder();
+                    output.Append(error);
+
+                    if(payload.RootElement.TryGetProperty("error_description", out var description))
+                    {
+                        output.Append(" ");
+                        output.Append(description);
+                    }
+                    return OAuthTokenResponse.Failed(new Exception(output.ToString()));
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.GitHub/Configuration/GithubHandler.cs:88]: It's quite a lot of dependant code to copy in, to get the custom error message. 

Did you consider a delegating http handler on the `Options.BackChannel` httpclient ?

Or perhaps it is worth opening an issue on ASP.NET to see about the possiblities of better error handling overrides?

Comment (jptissot) [reply src/OrchardCore.Modules/OrchardCore.GitHub/Configuration/GithubHandler.cs:88]: I agree with you, too much dependant code to copy. 

I think ASP.NET should have an extensibility point to parse and validate the OAuth response.

**Comment by sebastienros** [comment]
Maybe file an issue on aspnet to ask for a better error message, keep copying the code for now, add a reference to the issue in the code, and when the issue is fixed we can remove the code. The comment can also mention to keep this code in sync with the one in aspnet.

**Comment by jptissot** [comment]
> Maybe file an issue on aspnet to ask for a better error message, keep copying the code for now, add a reference to the issue in the code, and when the issue is fixed we can remove the code. The comment can also mention to keep this code in sync with the one in aspnet.

Done

**Pull request merged**